src/coredump/coredump-register.c

```diff
@@ -0,0 +1,160 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include "coredump-register.h"
+#include "errno-util.h"
+#include "fd-util.h"
+#include "json-util.h"
+#include "log.h"
+#include "pidfd-util.h"
+#include "process-util.h"
+#include "string-util.h"
+#include "sysctl-util.h"
+#include "varlink-io.systemd.Coredump.Register.h"
+#include "varlink-util.h"
+
+static int set_core_pattern(const char *prefix, const char *val) {
+        int r;
+
+        assert(val);
+
+        _cleanup_free_ char *buf = NULL;
+        if (prefix) {
+                buf = strjoin(prefix, val);
+                if (!buf)
+                        return -ENOMEM;
+                val = buf;
+        }
+
+        /* Since be1e0283021ec73c2eb92839db9a471a068709d9 (v6.17), which is backported as
+         * 7d7c1fb85cba5627bbe741fb7539c709435e3848 (v6.16.8), the kernel accepts any invalid patterns. The
+         * written pattern is checked only on read, spuriously... Let's first save the original value, then
+         * try to write the requested patter, and validate by reading the value. If the validation failed,
+         * let's revert to the original value. The spurious behavior is fixed by
+         * a779e27f24aeb679969ddd1fdd7f636e22ddbc1e (v6.18). */
+
+        _cleanup_free_ char *original = NULL;
+        r = sysctl_read("kernel/core_pattern", &original);
+        if (r >= 0 && streq(original, val))
+                return 0; /* Already set. */
+
+        r = sysctl_write("kernel/core_pattern", val);
+        if (r >= 0) {
+                _cleanup_free_ char *current = NULL;
+                r = sysctl_read("kernel/core_pattern", &current);
+                if (r >= 0 && streq(current, val))
+                        return 0; /* Yay! */
+        }
```

Comment:
hmm, this issue came up with other sysctls too, don't remember which one though. But let's maybe turn this into a proper helper for the future? i.e. introduce a helper sysctl_write_verify() that does just this logic

because i am pretty sure this will come up again and again sooner or later, and next time we should really have the helpers ready then